src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,146 @@
+        lastNode = node;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        checkIndexInBounds(index);
+        Node<T> nextNode = findByIndex(index);
+        Node<T> node = new Node<T>(nextNode.prev, value, nextNode);
+        if (index == 0) {
+            firstNode = node;
+        } else {
+            nextNode.prev.next = node;
+        }
+        nextNode.prev = node;
+        size++;
+        return;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexInBounds(index);
+        return findByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexInBounds(index);
+        Node<T> desiredNode = findByIndex(index);
+        T oldValue = desiredNode.value;
+        desiredNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexInBounds(index);
+        Node<T> desiredNode = findByIndex(index);
+        unlink(desiredNode);
+        T oldValue = desiredNode.value;
+        size--;
+        return oldValue;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = firstNode;
+        int i = 0;
+        while (currentNode != null) {
+            if ((currentNode.value != null && currentNode.value.equals(object))
+                    || currentNode.value == object) {
+                unlink(currentNode);
+                size--;
+                return true;
+            }
+            currentNode = currentNode.next;
+            i++;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndexInBounds(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index " + index + " is out of bounds " + size);
+        }
+    }
+
+    private void unlink(Node<T> node) {
+        if (node.prev != null) {
+            node.prev.next = node.next;
+        } else {
+            firstNode = node.next;
+        }
+        if (node.next != null) {
+            node.next.prev = node.prev;
+        } else {
+            lastNode = node.prev;
+        }
+    }
+
+    private Node<T> findByIndex(int index) {
+        Node<T> currentNode = firstNode;
+        int i = 0;
+        while (currentNode != null) {
+            if (i == index) {
+                return currentNode;
+            }
+            currentNode = currentNode.next;
+            i++;
+        }
+        return null;
+    }
```

Comment:
you should iterate from the closest end of the list, as indicated here:
![image](https://github.com/mate-academy/jv-linked-list/assets/49767049/d282ce4e-6a2c-451f-91c8-8932af5f0410)